### PR TITLE
[WIP] Add JSON writer to SimplerKernelTimer

### PR DESCRIPTION
**Pull request by dalg24**
cc @pgrete

### CODE REVIEW

**Comment by DavidPoliakoff** [comment]
I like the approach. For anybody reading, the approach is that we're adding a converter from the `simple-kernel-timer` binary format to a JSON one. This should over time remove the need for two highly similar code bases. More review for tomorrow, but a good start

File: profiling/simple-kernel-timer/kp_json_writer.cpp [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+			}
+		}
+
+		fclose(the_file);
+	}
+
+	std::sort(kernelInfo.begin(), kernelInfo.end(), compareKernelPerformanceInfo);
+
+	for(int i = 0; i < kernelInfo.size(); i++) {
+    if(kernelInfo[i]->getKernelType() != REGION) {
```
Comment (aprokop) [inline profiling/simple-kernel-timer/kp_json_writer.cpp:165]: mix of tab and spaces in the code

**Comment by pgrete** [comment]
Thanks @dalg24 
I didn't realize the support for regions within the simple kernel timer was already available and that a fix was as simple as adding a new writer.
I only tested the `simple-kernel-timer-json`  which does not include output for regions.

With respect to the output format of the json file, how much work do you anticipate it'd be to include the information about the nested structure of regions (similar to the space-time-stack)?

**Comment by dalg24** [comment]
@pgrete How would you like the json file to look like to capture the nested information?

**Comment by pgrete** [comment]
Something like the top-down time tree of the `space-time-stack` would be closest to the way we intent to parse the information.
(Our goal is to setup continuous benchmarking, and collect and compare the "application profiles" [i.e., how much time is spent in potentially nested regions and kernels for a canonical test problem])

**Comment by jrmadsen** [comment]
> With respect to the output format of the json file, how much work do you anticipate it'd be to include the information about the nested structure of regions (similar to the space-time-stack)?

So this sort of feature is supported by "external" tools. The default output format for the [timemory toolkit](https://github.com/NERSC/timemory) being developed at NERSC is JSON and will essentially give you the same JSON schema for all the different measurements (timers, memory, hw counters, etc. -- the only real differences in the output schemas are some report scalars (e.g. timers), others arrays (hw counters). The main library will let you dynamically select the measurement types but there are a bunch of zero-config libs installed to `<prefix>/lib/timemory/kokkos-tools/` folder depending on which features you enable.

**Comment by jrmadsen** [comment]
@pgrete quick start if you are interested. As far as setting up that continuous benchmarking, the library is designed for working with the actual numerical data in C++ and Python. You can create your own measurement types and :

```console
$ git clone https://github.com/NERSC/timemory.git timemory && cd timemory
$ python setup.py install --user --enable-{mpi,cuda,cupti,kokkos} -- -DTIMEMORY_BUILD_KOKKOS_CONFIG=ON
```

You need the `-DTIMEMORY_BUILD_KOKKOS_CONFIG=ON` to build the zero-config libs and on my Linux system, this resulted in:

```console
$ ls ~/.local/lib/timemory/kokkos-tools/
kp_timemory_context_switch.so    kp_timemory_cuda_profiler.so  kp_timemory_gpu_cache_misses.so  kp_timemory_gpu_instructions.so  kp_timemory_gpu_single_flops.so  kp_timemory_memory.so         kp_timemory_trip_count.so
kp_timemory_cuda_activity.so     kp_timemory_filter.so         kp_timemory_gpu_double_flops.so  kp_timemory_gpu_ipc.so           kp_timemory_gpu_stalls.so        kp_timemory.so
kp_timemory_cuda_event.so        kp_timemory_gperftools.so     kp_timemory_gpu_efficiency.so    kp_timemory_gpu_l2_total.so      kp_timemory_gpu_warps.so         kp_timemory_timers_memory.so
kp_timemory_cuda_hw_counters.so  kp_timemory_gpu_branching.so  kp_timemory_gpu_half_flops.so    kp_timemory_gpu_roofline.so      kp_timemory_io.so                kp_timemory_timers.so
```

just set `KOKKOS_PROFILE_LIBRARY=/.../kp_timemory_timers.so` and which a little python script to print it out "pretty":

```console
$ ./pretty_print_json.py timemory-kp-timemory-timers-output/2020-11-25_05.34_PM/wall.json
{
    "timemory": {
        "wall_clock": {
            "num_ranks": 1,
            "concurrency": 1,
            "type": "wall",
            "description": "Real-clock timer (i.e. wall-clock timer)",
            "unit_value": 1,
            "unit_repr": "sec",
            "properties": {
                "cereal_class_version": 0,
                "value": 70,
                "enum": "WALL_CLOCK",
                "id": "wall_clock",
                "ids": [
                    "real_clock",
                    "virtual_clock",
                    "wall_clock"
                ]
            },
            "ranks": [
                {
                    "rank": 0,
                    "graph_size": 10,
                    "graph": [
                        {
                            "hash": 16594354393280032487,
                            "prefix": ">>> kokkos/allocate/Host/y",
                            "depth": 0,
                            "entry": {
                                "cereal_class_version": 0,
                                "is_transient": true,
                                "laps": 1,
                                "value": 109825440027,
                                "accum": 109825440027,
                                "last": {},
                                "repr_data": 109.825440027,
                                "repr_display": 109.825440027
                            },
                            "stats": {
                                "cereal_class_version": 0,
                                "sum": 109.825440027,
                                "count": 1,
                                "min": 109.825440027,
                                "max": 109.825440027,
                                "sqr": 12061.627277124173,
                                "mean": 109.825440027,
                                "stddev": 0.0
                            },
                            "rolling_hash": 16594354393280032487
                        },
                        {
                            "hash": 8260192932196939784,
                            "prefix": ">>> |_kokkos/dev0/Kokkos::View::initialization",
                            "depth": 1,
                            "entry": {
```

There is also "*.tree.json" format also which is hierarchical (as opposed to the flat array above), includes exclusive calculations, and designed to be compatible with a LLNL project called Hatchet which converts it to pandas dataframes.

**Review by DavidPoliakoff — APPROVED** [review]

**Pull request merged**